# Стажировка





# Задача: Исследовать датасет

*Для этого нужно подключить библиотеку Numpy.



In [ ]:
# Загрузка модуля pandas
import pandas as pd

Затем нужно создать датасет. Для этого нужно загрузить датасет в Colab.

In [ ]:
# Загрузка файла из облака в colab
import gdown
gdown.download('https://storage.yandexcloud.net/terratraineeship/23_InspectrumClinic_RS/datasets/%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%20%D0%BF%D1%80%D0%BE%D1%84%D0%BF%D0%B0%D1%82%D0%BE%D0%BB%D0%BE%D0%B3%20%D1%87%D0%B0%D1%81%D1%82%D1%8C.xlsx', None, quiet=True)

'%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%20%D0%BF%D1%80%D0%BE%D1%84%D0%BF%D0%B0%D1%82%D0%BE%D0%BB%D0%BE%D0%B3%20%D1%87%D0%B0%D1%81%D1%82%D1%8C.xlsx'

Создадаем переменную data, в которую с помощью функции pd.read_excel() загрузим данные из файла profpatolog.xlsx:

**Переименуем**

In [ ]:
import os

# Предположим, что файл загружен с именем "old_filename.txt"
old_filename = "%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%20%D0%BF%D1%80%D0%BE%D1%84%D0%BF%D0%B0%D1%82%D0%BE%D0%BB%D0%BE%D0%B3%20%D1%87%D0%B0%D1%81%D1%82%D1%8C.xlsx"

# Получение нового имени файла
new_filename = "profpatolog.xlsx"

# Переименование файла
os.rename(old_filename, new_filename)


In [ ]:
data = pd.read_excel('profpatolog.xlsx')


Смотрим что из себя представляет датасет.

In [ ]:
data

Выведем первые 5 строк из таблицы.

In [ ]:
data.head()

ДокументПрохождениеМедосмотра                                Клиент  \
0  7f41ca7e-7dac-11ec-80fc-0cc47aab8067  c91e3407-b81f-11e9-80cf-0cc47aab8067   
1  2ee1f9ac-83dd-11ec-80fc-0cc47aab8067  1a1a5fea-8b7f-11eb-80e8-0cc47aab8067   
2  9cbb191e-8e24-11ec-80fe-0cc47aab8067  7fe324e0-fa51-11eb-80f6-0cc47aab8067   
3  b8a33910-8eec-11ec-80fe-0cc47aab8067  98f9d761-8d65-11ec-80fe-0cc47aab8067   
4  0bb0d181-938c-11ec-80fe-0cc47aab8067  009aac7d-7bcf-11eb-80e8-0cc47aab8067   

  КлиентДатаРождения КлиентПол КлиентПолПредставление  \
0         1976-10-24   Женский                Женский   
1         1987-11-18   Мужской                Мужской   
2         1985-09-08   Мужской                Мужской   
3         1988-04-25   Мужской                Мужской   
4         1984-04-17   Мужской                Мужской   

  ДатаЗавершенияМедосмотра  ТипМедосмотра ТипМедосмотраПредставление  \
0               2022-01-25  Периодический              Периодический   
1               2022-02-02  Периодический              Периодический   
2               2022-02-15  Периодический              Периодический   
3               2022-02-16  Периодический              Периодический   
4               2022-02-22  Периодический              Периодический   

   ПсихОсвидетельствование                            Контрагент  ...  \
0                    False  23c173a7-30d9-11e7-80d8-708bcd7f83cd  ...   
1                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
2                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
3                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
4                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   

  Оториноларингология3_Аудиометрия1_Значение  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
0                        NaN                                     NaN   
1                        NaN                                     NaN   
2                        NaN                                     NaN   
3                        NaN                                     NaN   
4                        NaN                                     NaN   

  Терапия2_ЗаключениеФОГК_Значение  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

  Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                              

Крайние 5 строк.

In [ ]:
data.tail()

Первично датасет готов к обработке.

# Анализ датасета

Проанализируем данные для нейронной сети.

In [ ]:
data.shape[1]

4587

Имеет 1982 строки и 4587 столбцов .

In [ ]:
# Сколько элементов в таблице
data.size 

9091434

In [ ]:
data.count()

ДокументПрохождениеМедосмотра                                             1982
Клиент                                                                    1982
КлиентДатаРождения                                                        1982
КлиентПол                                                                 1982
КлиентПолПредставление                                                    1982
                                                                          ... 
Стоматология2_ВП2_ЗначениеПредставление                                      1
Терапия2_ЗаключениеФОГК_Значение                                             1
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                                1
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                    1
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление       1
Length: 4587, dtype: int64

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1982 entries, 0 to 1981
Columns: 4587 entries, ДокументПрохождениеМедосмотра to Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление
dtypes: bool(1), datetime64[ns](43), float64(1253), object(3290)
memory usage: 69.3+ MB


DataFrame состоит из 1982 записей (строк) и 4587 столбцов.

Индекс (индексная колонка) имеет тип RangeIndex и содержит значения от 0 до 1981.
В DataFrame присутствуют различные типы данных:
1 cтолбец имеет тип bool.

43 столбца имеют тип datetime64[ns], что указывает на наличие дат и времени.

1253 столбца имеют тип float64, что указывает на наличие числовых значений с плавающей запятой.

3290 столбцов имеют тип object, что обычно означает наличие строковых данных или смешанных типов данных.

Общий объем памяти, используемый для хранения DataFrame, составляет примерно 69,3 мегабайта.

In [ ]:
#Смотрим есть ли пустые ячейки
data.isna().head()

ДокументПрохождениеМедосмотра  Клиент  КлиентДатаРождения  КлиентПол  \
0                          False   False               False      False   
1                          False   False               False      False   
2                          False   False               False      False   
3                          False   False               False      False   
4                          False   False               False      False   

   КлиентПолПредставление  ДатаЗавершенияМедосмотра  ТипМедосмотра  \
0                   False                     False          False   
1                   False                     False          False   
2                   False                     False          False   
3                   False                     False          False   
4                   False                     False          False   

   ТипМедосмотраПредставление  ПсихОсвидетельствование  Контрагент  ...  \
0                       False                    False       False  ...   
1                       False                    False       False  ...   
2                       False                    False       False  ...   
3                       False                    False       False  ...   
4                       False                    False       False  ...   

   Оториноларингология3_Аудиометрия1_Значение  \
0                                        True   
1                                        True   
2                                        True   
3                                        True   
4                                        True   

   Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                               True         
1                                               True         
2                                               True         
3                                               True         
4                                               True         

   Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            True   
1                                            True   
2                                            True   
3                                            True   
4                                            True   

   Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                               True             
1                                               True             
2                                               True             
3                                               True             
4                                               True             

   Стоматология2_ВП2_Значение  Стоматология2_ВП2_ЗначениеПредставление  \
0                        True                                     True   
1                        True                                     True   
2                        True                                     True   
3                        True                                     True   
4                        True                                     True   

   Терапия2_ЗаключениеФОГК_Значение  \
0                              True   
1                              True   
2                              True   
3                              True   
4                              True   

   Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           True   
1                                           True   
2                                           True   
3                                           True   
4                                           True   

   Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение  \
0                                               True           
1                                               True           
2                                               True           
3                                              

In [ ]:
#Сколько пустых ячеек?
data.isna().sum()

ДокументПрохождениеМедосмотра                                                0
Клиент                                                                       0
КлиентДатаРождения                                                           0
КлиентПол                                                                    0
КлиентПолПредставление                                                       0
                                                                          ... 
Стоматология2_ВП2_ЗначениеПредставление                                   1981
Терапия2_ЗаключениеФОГК_Значение                                          1981
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                             1981
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                 1981
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление    1981
Length: 4587, dtype: int64

Вообщем и целом видим какое-то количество пропусков есть.

In [ ]:
#Доля пропусков от общего количества элементов
data.isna().sum() / data.shape[0]

ДокументПрохождениеМедосмотра                                             0.000000
Клиент                                                                    0.000000
КлиентДатаРождения                                                        0.000000
КлиентПол                                                                 0.000000
КлиентПолПредставление                                                    0.000000
                                                                            ...   
Стоматология2_ВП2_ЗначениеПредставление                                   0.999495
Терапия2_ЗаключениеФОГК_Значение                                          0.999495
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                             0.999495
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                 0.999495
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление    0.999495
Length: 4587, dtype: float64

In [ ]:
# Общая сумма пропусков
data.isna().sum().sum()

8106978

In [ ]:
# Общий процент пропусков
proc = data.isna().sum().sum() / data.size
print('{:.1%}'.format(proc))

89.2%


Исходя из этих данных получается что пропущено 90% строк?

Параметр include=['O'] метода .describe() выведет статистику для строковых данных:

*   count - количество заполненных элементов
*   unique - количество уникальных элементов
*   top - самый распространенный элемент
*   freq - сколько раз встречается самый распространенный элемент


In [ ]:
# Вывод статистики для строковых данных.
data.describe(include=['O'])

ДокументПрохождениеМедосмотра  \
count                                   1982   
unique                                  1915   
top     e334cb0e-55a6-11ed-8108-0cc47aab8067   
freq                                       2   

                                      Клиент КлиентПол КлиентПолПредставление  \
count                                   1982      1982                   1982   
unique                                  1911         2                      2   
top     116fe6a1-455f-11ed-8107-0cc47aab8067   Мужской                Мужской   
freq                                       3      1451                   1451   

        ТипМедосмотра ТипМедосмотраПредставление  \
count            1982                       1982   
unique              2                          2   
top     Периодический              Периодический   
freq             1981                       1981   

                                  Контрагент         Цех Профессия  \
count                                   1982         802      1982   
unique                                   177         183       690   
top     173d6ba7-12fe-11e7-80d3-708bcd7f83cd  Без отдела  Водитель   
freq                                     162          63        55   

       ЗаключениеМК  ... Оториноларингология3_Аудиометрия1_Значение  \
count          1575  ...                                          1   
unique            9  ...                                          1   
top           Годен  ...       2934f2d1-8e1f-11ec-80fe-0cc47aab8067   
freq           1434  ...                                          1   

       Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
count                                                   1        
unique                                                  1        
top                              не подлежит обследованию        
freq                                                    1        

       Оториноларингология1_ЖалобыДополнение_Значение  \
count                                               1   
unique                                              1   
top                                   на ком в горле.   
freq                                                1   

       Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
count                                                   1            
unique                                                  1            
top                                       на ком в горле.            
freq                                                    1            

       Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
count                           1                                       1   
unique                          1                                       1   
top                             C                                       C   
freq                            1                                       1   

                         Терапия2_ЗаключениеФОГК_Значение  \
count                                                   1   
unique                                                  1   
top     Рекомендации: лекарственные препараты принимат...   
freq                                                    1   

            Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
count                                                   1   
unique                                                  1   
top     Рекомендации: лекарственные препараты принимат...   
freq                                                    1   

       Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение  \
count                                                   1          
unique                                                  1          
top           девиация левого глаза к носу до 25 градусов          
freq                                                    1          

       Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление  
count        

**Выводы сделанные на основе анализа датасета**

*   Имеет 1982 строки и 4587 столбцов .
*   Общие данные после команды data.inf()
*   Почему-то общий процент пропусков 90%.
*   Есть уникальные идентификаторы. Они не нужны.
*   Есть закономерности, такие как пол влияет на прохождение врачей.
*   Нас иньересуют только 3 класса-**годен,ограниченно годен,негоден**.
*   Если не прошел психОсведетельствование это **Негоден**.


# Создание выборок

In [ ]:
# Создадим копию, так как будем редактировать таблицу.
data_backup = data.copy()
data.head()

ДокументПрохождениеМедосмотра                                Клиент  \
0  7f41ca7e-7dac-11ec-80fc-0cc47aab8067  c91e3407-b81f-11e9-80cf-0cc47aab8067   
1  2ee1f9ac-83dd-11ec-80fc-0cc47aab8067  1a1a5fea-8b7f-11eb-80e8-0cc47aab8067   
2  9cbb191e-8e24-11ec-80fe-0cc47aab8067  7fe324e0-fa51-11eb-80f6-0cc47aab8067   
3  b8a33910-8eec-11ec-80fe-0cc47aab8067  98f9d761-8d65-11ec-80fe-0cc47aab8067   
4  0bb0d181-938c-11ec-80fe-0cc47aab8067  009aac7d-7bcf-11eb-80e8-0cc47aab8067   

  КлиентДатаРождения КлиентПол КлиентПолПредставление  \
0         1976-10-24   Женский                Женский   
1         1987-11-18   Мужской                Мужской   
2         1985-09-08   Мужской                Мужской   
3         1988-04-25   Мужской                Мужской   
4         1984-04-17   Мужской                Мужской   

  ДатаЗавершенияМедосмотра  ТипМедосмотра ТипМедосмотраПредставление  \
0               2022-01-25  Периодический              Периодический   
1               2022-02-02  Периодический              Периодический   
2               2022-02-15  Периодический              Периодический   
3               2022-02-16  Периодический              Периодический   
4               2022-02-22  Периодический              Периодический   

   ПсихОсвидетельствование                            Контрагент  ...  \
0                    False  23c173a7-30d9-11e7-80d8-708bcd7f83cd  ...   
1                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
2                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
3                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
4                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   

  Оториноларингология3_Аудиометрия1_Значение  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
0                        NaN                                     NaN   
1                        NaN                                     NaN   
2                        NaN                                     NaN   
3                        NaN                                     NaN   
4                        NaN                                     NaN   

  Терапия2_ЗаключениеФОГК_Значение  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

  Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                              

In [ ]:
#Восстановление датасета.
data = data_backup.copy()
data.head()

Хочу удалить пустые столбцы.

In [ ]:
import pandas as pd

# Удаление столбцов, в которых все значения являются пустыми
data_without_empty_columns = data.dropna(axis=1, how='all')

# Удаление строк, в которых все значения являются пустыми
data_cleaned = data_without_empty_columns.dropna(axis=0, how='all')

# В результате data_cleaned будет содержать данные без пустых строк и столбцов, в которых все значения являются пустыми


In [ ]:
# Общий процент пропусков
proc = data_cleaned.isna().sum().sum() / data.size
print('{:.1%}'.format(proc))

69.9%


Посмотрим общий процент заполненности строк. Может поможет.

In [ ]:

# Получение общего количества столбцов
total_columns = len(data_cleaned.columns)

# Инициализация переменных для подсчета суммы процентов заполненности
total_filled_percentage = 0
total_rows = 0

# Проход по каждой строке
for index, row in data_cleaned.iterrows():
    # Подсчет количества непустых значений в строке
    filled_values = row.count()
    
    # Подсчет процента заполненности строки
    filled_percentage = (filled_values / total_columns) * 100
    
    # Обновление суммы процентов заполненности и количества строк
    total_filled_percentage += filled_percentage
    total_rows += 1

# Вычисление среднего процента заполненности строк
average_filled_percentage = total_filled_percentage / total_rows

print("Средний процент заполненности строк:", average_filled_percentage)


Средний процент заполненности строк: 13.413402229572085


In [ ]:
data_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1982 entries, 0 to 1981
Columns: 3703 entries, ДокументПрохождениеМедосмотра to Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление
dtypes: bool(1), datetime64[ns](43), float64(369), object(3290)
memory usage: 56.0+ MB


Уиды из 1С содержат по 36 символов, попробуем удалить их вместе со столбцами.

**Удалю ненужные столбцы.**

In [ ]:
# Указать список столбцов, которые нужно удалить
columns_to_drop = ['ДокументПрохождениеМедосмотра', 'Клиент', 'Контрагент']

# Удалить указанные столбцы из датасета
data_cleaned.drop(columns=columns_to_drop, inplace=True)


In [ ]:
data_cleaned.head()

Нужно узнать сколько и каких типов данных в столбце ЗаключениеМК, так как эти данные нужно поделить на 3 класса и использовать их затем для обучения нейронки.

In [ ]:
data_cleaned['ЗаключениеМК'].value_counts()


Годен                    1434
Годен с ограничениями     126
Негоден                    11
Name: ЗаключениеМК, dtype: int64

Обнаружен машинист бульдозера и газоспасатель! Причем один Газоспасатель неуёмный оказался и не хотел удаляться.


In [ ]:
data_cleaned = data_cleaned[~data_cleaned['ЗаключениеМК'].isin(['машинист бульдозера', 'Газоспасатель', ' Газоспасатель'])]


Нужно оставшееся превести к 3 значениям. Годен, Годен с ограничениями, Негоден.

In [ ]:
data_cleaned['ЗаключениеМК'] = data_cleaned['ЗаключениеМК'].replace({
    'ГоденСКоррекциейЗрения': 'Годен с ограничениями',
    'ОграниченноГоден': 'Годен с ограничениями',
    'ГоденБезРаботНаВысотах': 'Годен с ограничениями',
    'ВременноНегоден': 'Негоден',
    'НуждаетсяВДообследованииИЛечении': 'Негоден'
})



**Получилось!**

Годен                    1434

Годен с ограничениями     126

Негоден                    11

Name: ЗаключениеМК, dtype: int64

Разбалансирован немного, но ничего. Можно пробовать разбить на выборки.

**Попробуем поколдовать при создании выборки для обучения.**

In [ ]:
from sklearn.model_selection import train_test_split

# Разделение данных на признаки (X) и целевую переменную (y)
X = data_cleaned.drop(columns=['ЗаключениеМК'])
y = data_cleaned['ЗаключениеМК']

# Разделение на обучающую выборку, валидационную выборку и тестовую выборку
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_test, y_test, test_size=0.5, random_state=42)


In [ ]:
# Проверка формы сформированных данных
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(1384, 3699) (1384,)
(297, 3699) (297,)


Создали, теперь токенизируем.